lib/deep_cover/analyser/ruby25_like_branch.rb

```diff
@@ -0,0 +1,214 @@
+        branches_locs[1] = extend_elsif_range(branches_locs[1])
+
+        clauses_infos = infos_for_branches(branches_locs, sub_keys, empty_fallback_locs, execution_counts: execution_counts, node_range: node_range)
+        [cond_info, clauses_infos]
+      end
+
+      def handle_short_circuit
+        cond_info = [operator, *node_loc_infos]
+        sub_keys = [:then, :else]
+        sub_keys.reverse! if node.is_a?(Node::Or)
+
+        [cond_info, infos_for_branches(branches, sub_keys, [nil, nil])]
+      end
+
+      def handle_until_while
+        key = loc_hash[:keyword].source.to_sym
+        base_info = [key, *node_loc_infos]
+        body_node = if node.is_a?(Node::WhilePost) || node.is_a?(Node::UntilPost)
+                      if body.instructions.present?
+                        end_pos = body.instructions.last.expression.end_pos
+                        body.instructions.first.expression.with(end_pos: end_pos)
+                      else
+                        body.loc_hash[:end].begin
+                      end
+                    elsif body.is_a?(Node::Begin) && node.body.expressions.present?
+                      end_pos = body.expressions.last.expression.end_pos
+                      body.expressions.first.expression.with(end_pos: end_pos)
+                    else
+                      body
+                    end
+
+        [base_info, {[:body, *node_loc_infos(body_node)] => body.execution_count}]
+      end
+
+      protected
+
+      # If the actual else clause (final one) of an if...elsif...end is empty, then Ruby makes the
+      # node reach the `end` in its branch coverage output
+      def extend_elsif_range(node = self.node)
+        if node.is_a?(Node::If) && node.style == :elsif
+          deepest_if = node.deepest_elsif_node || node
+          if deepest_if.false_branch.is_a?(Node::EmptyBody)
+            return node.expression.with(end_pos: node.root_if_node.loc_hash[:end].begin_pos)
+          end
+        end
+        node
+      end
+
+      def infos_for_branch(branch, key, empty_fallback_loc, execution_count: nil, node_range: node)
+        if !branch.is_a?(Node::EmptyBody)
+          loc = branch
+        elsif branch.expression
+          # There is clause, but it is empty
+          loc = empty_fallback_loc
+        else
+          # There is no clause
+          loc = node_range
+        end
+
+        execution_count ||= branch.execution_count
+        [[key, *node_loc_infos(loc)], execution_count]
+      end
+
+      def infos_for_branches(branches, keys, empty_fallback_locs, execution_counts: [], node_range: node)
+        branches_infos = branches.map.with_index do |branch, i|
+          infos_for_branch(branch, keys[i], empty_fallback_locs[i], execution_count: execution_counts[i], node_range: node_range)
+        end
+        branches_infos.to_h
+      end
+
+      def node_loc_infos(source_range = node)
+        if source_range.is_a?(Node)
```

Comment:
Maybe hairpicking, but why not have two methods like `node_loc_infos` and `loc_infos` instead of one that takes either a `node` or a `source_location`? Or at least consider renaming the argument `node_or_loc` or similar...

Reply:
Because the method is called from `infos_for_branch`, which is called from `infos_for_branches`. So if I have 2 methods, I need to transfer this logic elsewhere, making things more complicated.